From 7198b17ef9e1db39e76ddbae0509c1058aa5c321 Mon Sep 17 00:00:00 2001
From: Hirokazu Hata <h.hata.ai.t@gmail.com>
Date: Sun, 31 Aug 2025 02:28:09 +0900
Subject: [PATCH] fix(view): adapt to the changes for
 "nvim_set_decoration_provider"

A recent commit in Neovim (5edbabdbec0) changed the decoration
provider API (vim.api.nvim_set_decoration_provider), replacing
the on_line callback with on_range.

This breaking change, which also alters the callback arguments,
causes an error when using trouble.nvim with recent Neovim nightly
builds. This commit updates the call in lua/trouble/view/treesitter.lua
to use the new on_range API, restoring compatibility.

 - Relevant Neovim Commit: neovim/neovim@5edbabdbec0
 - Relevant Neovim PR: https://github.com/neovim/neovim/pull/31400
---
 lua/trouble/view/treesitter.lua | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/lua/trouble/view/treesitter.lua b/lua/trouble/view/treesitter.lua
index 1fd35fc..8f6fbea 100644
--- a/lua/trouble/view/treesitter.lua
+++ b/lua/trouble/view/treesitter.lua
@@ -29,10 +29,18 @@ function M.setup()
   end
   M.did_setup = true
 
-  vim.api.nvim_set_decoration_provider(ns, {
-    on_win = wrap("_on_win"),
-    on_line = wrap("_on_line"),
-  })
+  -- https://github.com/neovim/neovim/commit/5edbabdbec0ac3fba33be8afc008845130158583
+  if vim.fn.has("nvim-0.12.0") == 1 then
+    vim.api.nvim_set_decoration_provider(ns, {
+      on_win = wrap("_on_win"),
+      on_range = wrap("_on_range"),
+    })
+  else
+    vim.api.nvim_set_decoration_provider(ns, {
+      on_win = wrap("_on_win"),
+      on_line = wrap("_on_line"),
+    })
+  end
 
   vim.api.nvim_create_autocmd("BufWipeout", {
     group = vim.api.nvim_create_augroup("trouble.treesitter.hl", { clear = true }),
-- 
2.50.1

